## Exercise 4.9 - Part 1

In [3]:
## Importing Libraries

In [4]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt
import seaborn as sns
import scipy

In [5]:
data_path=r'/Users/Lilian/Downloads/Data Analytics Course/Data Analytics Immersion/Acheivement 4/04-2024 Instacart Basket Analysis/'

In [6]:
df_customers = pd.read_csv(os.path.join(data_path,'02 Data','Original Data','customers.csv'))

In [7]:
df_customers.shape

(206209, 10)

In [8]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
df_customers.rename(columns={'First Name': 'first_name'}, inplace=True)

In [10]:
df_customers.rename(columns={'Surnam': 'last_Name'}, inplace=True)

In [11]:
df_customers.rename(columns={'fam_status': 'marital_status'}, inplace=True)

In [12]:
df_customers.rename(columns={'n_dependants': 'number_of_fam_members'}, inplace=True)

In [13]:
df_customers.rename(columns={'Age': 'age'}, inplace=True)

In [14]:
df_customers.rename(columns={'STATE': 'state'}, inplace=True)

In [15]:
df_customers.rename(columns={'Gender': 'gender'}, inplace=True)

In [16]:
df_customers.head()

user_id first_name last_Name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_fam_members marital_status  income  
0                      3        married  165665  
1                      0         single   59285  
2                      2        married   99568  
3                      0         single   42049  
4                      1        married   40374

In [17]:
df_customers['income'].describe()

count    206209.000000
mean      94632.852548
std       42473.786988
min       25903.000000
25%       59874.000000
50%       93547.000000
75%      124244.000000
max      593901.000000
Name: income, dtype: float64

In [18]:
df_customers['age'].describe()

count    206209.000000
mean         49.501646
std          18.480962
min          18.000000
25%          33.000000
50%          49.000000
75%          66.000000
max          81.000000
Name: age, dtype: float64

In [19]:
# finding missing values
df_customers.isnull().sum()

user_id                      0
first_name               11259
last_Name                    0
gender                       0
state                        0
age                          0
date_joined                  0
number_of_fam_members        0
marital_status               0
income                       0
dtype: int64

#### first_name column has 11259 null values, but probably not affecting the main analysis - can be left empty 

In [20]:
#checking for duplicated values and setting them into a new df
df_customers_dups = df_customers[df_customers.duplicated()]

In [21]:
df_customers_dups.shape

(0, 10)

#### no duplicates found

### checking for mixed type data 

In [22]:
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].map(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)

first_name


In [105]:
df_customers['first_name']

0          Deborah
1         Patricia
2          Kenneth
3         Michelle
4              Ann
            ...   
206204        Lisa
206205      Jeremy
206206       Doris
206207        Rose
206208     Cynthia
Name: first_name, Length: 206209, dtype: object

In [23]:
# converting all values within the first name column to string
df_customers['first_name'] = df_customers['first_name'].astype('str')

In [24]:
for col in df_customers.columns.tolist():
  weird = (df_customers[[col]].map(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_customers[weird]) > 0:
    print (col)
else:
    print ("no more mixed data")

no more mixed data


In [25]:
df_customers.dtypes

user_id                   int64
first_name               object
last_Name                object
gender                   object
state                    object
age                       int64
date_joined              object
number_of_fam_members     int64
marital_status           object
income                    int64
dtype: object

### Importing rest of instacard data

In [26]:
orders_products_merge = pd.read_pickle(os.path.join(data_path,'02 Data','Prepared Data','orders_products_merge_Grouped_Agged.pkl'))

In [27]:
orders_products_merge.columns

Index(['product_id', 'product_name', 'aisle_id', 'department_id', 'prices',
       'order_id', 'user_id', 'eval_set', 'order_number', 'order_dow',
       'order_hour_of_day', 'days_since_prior_order', 'add_to_cart_order',
       'reordered', '_merge', 'busiest_days', 'busiest_period_of_day',
       'max_order', 'loyalty_flag', 'average_price_spent', 'Spending_flag',
       'ordering_median', 'Customer_order_frequency'],
      dtype='object')

In [28]:
# drop _merge column before merging with customer data
orders_products_merge = orders_products_merge.drop(columns=['_merge'])

In [29]:
# merging product_order and customer datasets into ords_prods_customers_merge
ords_prods_customers_merge = orders_products_merge.merge(df_customers, on = 'user_id', indicator = True)

In [30]:
ords_prods_customers_merge.head() 

product_id                product_name  aisle_id  department_id  prices  \
0           1  Chocolate Sandwich Cookies        61             19     5.8   
1           1  Chocolate Sandwich Cookies        61             19     5.8   
2         907        Premium Sliced Bacon       106             12    20.0   
3         907        Premium Sliced Bacon       106             12    20.0   
4        1000                    Apricots        18             10    12.9   

   order_id  user_id eval_set  order_number  order_dow  ...  first_name  \
0   3139998      138    prior            28          6  ...     Charles   
1   1977647      138    prior            30          6  ...     Charles   
2   3160996      138    prior             1          5  ...     Charles   
3   2254091      138    prior            10          5  ...     Charles   
4    505689      138    prior             9          6  ...     Charles   

   last_Name  gender      state age date_joined  number_of_fam_members  \
0        Cox    Male  Minnesota  81    8/1/2019                      1   
1        Cox    Male  Minnesota  81    8/1/2019                      1   
2        Cox    Male  Minnesota  81    8/1/2019                      1   
3        Cox    Male  Minnesota  81    8/1/2019                      1   
4        Cox    Male  Minnesota  81    8/1/2019                      1   

  marital_status  income _merge  
0        married   49620   both  
1        married   49620   both  
2        married   49620   both  
3        married   49620   both  
4        married   49620   both  

[5 rows x 32 columns]

In [32]:
#exporting merged dataframes : ords_prods_customers_merge

ords_prods_customers_merge.to_pickle(os.path.join(data_path, '02 Data','Prepared Data', 'orders_products_customers.pkl'))


